src/mdns/mdns_avahi.cpp

```diff
@@ -485,18 +485,24 @@ void PublisherAvahi::CallHostOrServiceCallback(AvahiEntryGroup *aGroup, otbrErro
 
     if ((serviceReg = FindServiceRegistration(aGroup)) != nullptr)
     {
-        serviceReg->Complete(aError);
-        if (aError != OTBR_ERROR_NONE)
+        if (aError == OTBR_ERROR_NONE)
         {
-            RemoveServiceRegistration(serviceReg->mName, serviceReg->mType);
+            serviceReg->Complete(aError);
+        }
+        else
+        {
+            RemoveServiceRegistration(serviceReg->mName, serviceReg->mType, aError);
         }
     }
     else if ((hostReg = FindHostRegistration(aGroup)) != nullptr)
     {
-        hostReg->Complete(aError);
-        if (aError != OTBR_ERROR_NONE)
+        if (aError == OTBR_ERROR_NONE)
+        {
+            hostReg->Complete(aError);
+        }
+        else
         {
-            RemoveHostRegistration(hostReg->mName);
+            RemoveHostRegistration(hostReg->mName, aError);
```

Comment:
nit: Add a new method for this logic? e.g. FinalizeHostRegistration(); ?

Reply:
`FinalizeHostRegistration` doesn't indicates whether the service is going to be removed or not.

Reply:
But deciding that we should remove the service is essentially part of the finalization logic?

The same logic pattern seems to occur in both mdns_avahi.cpp and  mdns_mdnssd.cpp. So, maybe better to encapsulate it?

Reply:
It sounds good, but we have to log specific mdnsResponder or avahi error when failed to register the service. so not easy to make a common function for it.